src/include/detail/linalg/matrix_with_ids.h

```diff
@@ -0,0 +1,209 @@
+      LayoutPolicy policy = LayoutPolicy()) noexcept
+    requires(std::is_same_v<LayoutPolicy, stdx::layout_right>)
+      : Base(nrows, ncols, policy)
+      , num_ids_(this->num_rows_)
+#ifdef __cpp_lib_smart_ptr_for_overwrite
+      , idsStorage_{std::make_unique_for_overwrite<IdsType[]>(this->num_rows_)}
+#else
+      , idsStorage_{new IdsType[this->num_rows_]}
+#endif
+  {
+  }
+
+  MatrixWithIds(
+      Base::size_type nrows,
+      Base::size_type ncols,
+      LayoutPolicy policy = LayoutPolicy()) noexcept
+    requires(std::is_same_v<LayoutPolicy, stdx::layout_left>)
+      : Base(nrows, ncols, policy)
+      , num_ids_(this->num_cols_)
+#ifdef __cpp_lib_smart_ptr_for_overwrite
+      , idsStorage_{std::make_unique_for_overwrite<IdsType[]>(this->num_cols_)}
+#else
+      , idsStorage_{new IdsType[this->num_cols_]}
+#endif
+  {
+  }
+
+  MatrixWithIds(
+      std::unique_ptr<T[]>&& storage,
+      std::unique_ptr<IdsType[]>&& ids_storage,
+      Base::size_type nrows,
+      Base::size_type ncols,
+      LayoutPolicy policy = LayoutPolicy()) noexcept
+      : Base(std::move(storage), nrows, ncols, policy)
+      , num_ids_{std::is_same<LayoutPolicy, stdx::layout_right>::value ? this->num_rows_ : this->num_cols_}
+      , idsStorage_{std::move(ids_storage)} {
+  }
+
+  /**
+   * Initializer list constructor. Useful for testing and for examples.
+   * The initializer list is assumed to be in row-major order.
+   */
+  MatrixWithIds(
+      std::initializer_list<std::initializer_list<T>> matrix,
+      std::initializer_list<T> ids) noexcept
+    requires(std::is_same_v<LayoutPolicy, stdx::layout_right>)
+      : Base(matrix)
+      , num_ids_(this->num_rows_)
+#ifdef __cpp_lib_smart_ptr_for_overwrite
+      , idsStorage_{std::make_unique_for_overwrite<IdsType[]>(this->num_rows_)}
+#else
+      , idsStorage_{new IdsType[this->num_rows_]}
+#endif
+  {
+    std::copy(ids.begin(), ids.end(), idsStorage_.get());
+  }
+
+  /**
+   * Initializer list constructor. Useful for testing and for examples.
+   * The initializer list is assumed to be in column-major order.
+   */
+  MatrixWithIds(
+      std::initializer_list<std::initializer_list<T>> matrix,
+      std::initializer_list<T> ids) noexcept
+    requires(std::is_same_v<LayoutPolicy, stdx::layout_left>)
+      : Base(matrix)
+      , num_ids_(this->num_cols_)
+#ifdef __cpp_lib_smart_ptr_for_overwrite
+      , idsStorage_{std::make_unique_for_overwrite<IdsType[]>(this->num_cols_)}
+#else
+      , idsStorage_{new IdsType[this->num_cols_]}
+#endif
+  {
+    std::copy(ids.begin(), ids.end(), idsStorage_.get());
+  }
+
+  size_t num_ids() const {
+    return num_ids_;
+  }
+
+  auto ids() {
+    return idsStorage_.get();
+  }
+
+  auto ids() const {
+    return idsStorage_.get();
+  }
+
+  auto raveledIds() {
+    return std::span(idsStorage_.get(), num_ids_);
+  }
+
+  auto raveledIds() const {
+    return std::span(idsStorage_.get(), num_ids_);
+  }
+
+  auto id(Base::index_type i) const {
```

Comment:
Don't mind, left it for now but I can remove if you have a preference either way.

Reply:
I tend to use the idiom of 
```
  auto ids = foo.ids();
  // ...
  auto x = ids[i];
```
rather than
```
  auto x = foo.id(i);
```
I use the former because in its use, it is more obvious with the ids are and what the costs for accessing them would be.  With `foo.id(i)` it is less clear when just reading `foo.id(i)` what is happening and what the costs are.

Let us start without it and see if that becomes a problem.

I think `foo.ids()[i]` should be "right out".

Reply:
Sure, removed.